feat: Deno Support

dcdunkan opened the pull request:
Brings Deno support to the module.

dcdunkan committed:
feat: Deno Support

dcdunkan committed:
ci: update jobs

dcdunkan committed:
ci: fix typo in key

dcdunkan committed:
ci: switch to npm for backporting

EdJoPaTo commented on package.json:
```diff
@@ -1,68 +1,35 @@
+    "node": ">=14"
+  },
+  "devDependencies": {
+    "@types/node": "^14.18.22",
```
What needs this dependency? The sources are completely JavaScript only and dont depend on Node.JS. Am I overlooking something?

dcdunkan replied:
No, it's my bad, sorry. Will remove it.

EdJoPaTo commented on package.json:
```diff
@@ -1,68 +1,35 @@
+  },
+  "scripts": {
+    "prepare": "npm run backport",
+    "backport": "deno2node tsconfig.json"
```
I think the backport script should be either defined in deno.jsonc or package.json but not both to prevent duplication? Adapting one might end up in different scripts later on?

The prepare script is relevant on installing from github directly so it might be a good idea to keep it in the package.json?

dcdunkan replied:
Well, I was following the setup that grammY uses. But yes, I guess it is better to keep it only in package.json.

EdJoPaTo committed:
test: move tests to source folder

specifying test/bla.test.ts is kinda pointless.
Move them to the sources and name them source/bla.test.ts is
not duplicating the test in the file path and is autodetected.

EdJoPaTo committed:
chore(gitignore): ignore npm pack output

npm pack is often used to check / test the npm package before publishing

EdJoPaTo committed:
style: format similar

EdJoPaTo commented on tsconfig.json:
```diff
@@ -1,11 +1,13 @@
+    "**/test.*"
+  ],
+  "compilerOptions": {
+    "rootDir": "source",
```
I think with my removal of the test folder this option is superfluous now? Maybe remove it to keep it simpler again?

dcdunkan replied:
Sure.

EdJoPaTo committed:
ci(deno): split lint and test

Both jobs can fail independently on their own as both could fail on
the same time.
Also rename to deno and pack after backport to see the result.

EdJoPaTo commented on .github/workflows/deno.yml:
```diff
@@ -0,0 +1,44 @@
+name: Deno
+
+on:
+  push:
+    branches: [main]
+  pull_request:
+    branches: [main]
```
The tests should always be run, not only on the main branch. Deno is rather fast so its not even taking that long and I prefer duplicate runs over missing runs. Other thoughts here?

```suggestion
  push:
  pull_request:
```

dcdunkan replied:
Sure. Still, I was only re-using the scripts/configs that I've been using before. But it is your repository and I agree with what you said.

dcdunkan committed:
chore: remove unwanted config option

as we've moved tests to source/

dcdunkan committed:
chore: remove unwanted dependency

dcdunkan committed:
chore: remove backport task

dcdunkan committed:
Update .github/workflows/deno.yml

Co-authored-by: EdJoPaTo <github@edjopato.de>

EdJoPaTo commented:
(as a response to a message inside a code comment) I expected them to be collected from grammY repos. Its not meant as criticism and more like "maybe we can improve and simplify things here". Maybe some grammY things could also be improved after rethinking some things here. (Or maybe I am just differently opinionated :P )
I probably want to have a base template for a typescript library like I currently have for node only. Having one with deno first and deno2node will probably simplify a lot of Node.js boilerplate stuff so its probably the way to go for me. Thats why I think a bit more about this migration to Deno in order to reuse what I learned here.

Maybe @KnorpelSenf or @wojpawlik (or others from the grammY / Deno bubble) have interesting feedback for a good minimal setup?

I am thinking about integrating things like [this](https://github.com/EdJoPaTo/deno-runtime-template/blob/main/deno.jsonc) in this PQ already.

Also I am thinking about removing the source folder to put the sources flat into the main folder. Not sure how well that works with deno2node?

wojpawlik commented on .github/workflows/deno.yml:
```diff
@@ -7,25 +7,28 @@ on:
 
-      - name: Test
-        run: deno test
+      - run: deno cache source/*.ts
+      - run: deno task check
+      - run: deno test
```
Redundant.

```suggestion
      - run: deno test
```

EdJoPaTo replied:
the benefit of splitting it up is easier visibility of errors on the CI interface. when the cache fails its seen directly. Also the time taken for each of the steps is easier distinguishable.

deno test doesn't do the checks anymore when not specified with --check stuff? also check checks not only the tests?

wojpawlik replied:
It checks tests and local files imported by them: https://deno.com/manual/getting_started/command_line_interface#type-checking-flags

EdJoPaTo replied:
But this would require to have every file included in some test?

wojpawlik replied:
Tests already import `mod.ts`.

EdJoPaTo replied:
in this case yeah :thinking: But in general I wouldnt be so sure about my testing efforts there :D

wojpawlik commented on package.json:
```diff
@@ -1,68 +1,35 @@
+    "url": "https://edjopato.de"
+  },
+  "scripts": {
+    "prepare": "npm run backport",
+    "backport": "deno2node tsconfig.json"
```
Unnecessary redirection.

```suggestion
    "prepare": "deno2node"
```

EdJoPaTo replied:
I think I do like the explicit "backport" script which is invoked by the prepare script. It's more describing. Omiting the tsconfig.json is probably a good thing to just use the default.

wojpawlik commented on .github/workflows/deno.yml:
```diff
@@ -2,9 +2,7 @@ name: Deno
 
 on:
   push:
-    branches: [main]
```
Without this line, if you make a branch in this repo and then make a PR, CI will run twice.

EdJoPaTo replied:
yeah. But with it creating a branch without a PR it will not run CI at all which is annoying too…

wojpawlik commented on .github/workflows/deno.yml:
```diff
@@ -0,0 +1,42 @@
+        run: npm install --ignore-scripts
+
+      - name: Backport
+        run: npm run backport
+
+      - run: npm pack
```
Redundant: `npm pack` invokes `"prepare"` script.

EdJoPaTo replied:
The idea is to get the output what is included in the package. Does the packing works? If not the publish will not work anyway. So I like this as a CI step.

wojpawlik commented on source/mod.ts:
```diff
@@ -0,0 +1,5 @@
+export * from "./types.ts";
+
+export * from "./html.ts";
+export * from "./markdown.ts";
+export * from "./markdownv2.ts";
```
Explicitness = easier navigation.
```suggestion
export type { Formatter } from "./types.ts";

export { html } from "./html.ts";
export { markdown } from "./markdown.ts";
export { markdownv2 } from "./markdownv2.ts";
```

EdJoPaTo replied:
true. On the other side it's nice to just export in the file instead of here too. 🤔

But I think this is independent from this PQ of providing Deno support.

wojpawlik commented on .github/workflows/deno.yml:
```diff
@@ -0,0 +1,42 @@
+          deno-version: v1.x
+      - uses: actions/checkout@v3
+
+      - run: deno lint
+      - run: deno fmt --check
```
They're _instant_, separate job is unnecessary.

EdJoPaTo replied:
but they can fail independently so when the fmt fails the tests can still run through and vice versa. Personally I think its annoying to push the fmt and then see it failing again because of tests directly after.

wojpawlik reviewed (commented):
> Also I am thinking about removing the source folder to put the sources flat into the main folder. Not sure how well that works with deno2node?

Too messy for my linking.

dcdunkan committed:
Apply suggestions from code review

Co-authored-by: Wojciech Pawlik <woj.pawlik@gmail.com>

EdJoPaTo committed:
chore(gitignore): ignore .vscode folder

EdJoPaTo committed:
build(npm): split backport/prepare script
